### ISSUE TITLE
rename has stale entries in the cache?

### ISSUE THREAD

**Issue by gourao** [bazil/fuse#119]
I am implementing a simple passthrough fuse driver to get acquainted with the bazil fuse implementation.

When I rename a file in my Rename handler, for about 10 seconds, I cannot stat the new file (the file I renamed the src to).

After a while, I can stat it.

LMK if you want me to point you at the source, but it's really a simple passthru driver.

**Comment by barakmich** [comment]
+1. I'm seeing this too. And trying to Invalidate the node locks up underlying fuse.

It appears to be caching the old nodes. 

Here's the debug output for `ls; mv nyc2.jpg nyc.jpg; ls`
```
 <- Getattr [ID=0x2 Node=0x1 Uid=1000 Gid=100 Pid=12861] 0x0 fl=0
 -> [ID=0x2] Getattr valid=1m0s ino=1 size=0 mode=dr-xr-xr-x
 <- Open [ID=0x3 Node=0x1 Uid=1000 Gid=100 Pid=12861] dir=true fl=OpenReadOnly+OpenDirectory+OpenNonblock
 -> [ID=0x3] Open 0x1 fl=0
 <- Read [ID=0x4 Node=0x1 Uid=1000 Gid=100 Pid=12861] 0x1 4096 @0x0 dir=true fl=0 lock=0 ffl=OpenReadOnly+OpenDirectory+OpenNonblock
 returning dirents: []fuse.Dirent{fuse.Dirent{Inode:0x0, Type:0x8, Name:"nyc2.jpg"}}
 -> [ID=0x4] Read 32
 <- Lookup [ID=0x5 Node=0x1 Uid=1000 Gid=100 Pid=12861] "nyc2.jpg"
stat {/nyc2.jpg}
stat {/nyc2.jpg}
 -> [ID=0x5] Lookup 0x2 gen=0 valid=1m0s attr={valid=1m0s ino=16273471275109000053 size=1774691 mode=-rw-rw-rw-}
 <- Getxattr [ID=0x6 Node=0x2 Uid=1000 Gid=100 Pid=12861] "security.capability" 20 @0
 -> [ID=0x6] Getxattr error=operation not supported
 <- Read [ID=0x7 Node=0x1 Uid=1000 Gid=100 Pid=12861] 0x1 4096 @0x20 dir=true fl=0 lock=0 ffl=OpenReadOnly+OpenDirectory+OpenNonblock
 -> [ID=0x7] Read 0
 <- Release [ID=0x8 Node=0x1 Uid=0 Gid=0 Pid=0] 0x1 fl=OpenReadOnly+OpenDirectory+OpenNonblock rfl=0 owner=0x0
 -> [ID=0x8] Release
 <- Lookup [ID=0x9 Node=0x1 Uid=1000 Gid=100 Pid=12873] "nyc.jpg"
stat {vol1 /nyc.jpg}
 -> [ID=0x9] Lookup error=ENOENT
 <- Lookup [ID=0xa Node=0x1 Uid=1000 Gid=100 Pid=12880] "nyc.jpg"
stat {vol1 /nyc.jpg}
 -> [ID=0xa] Lookup error=ENOENT
 <- Lookup [ID=0xb Node=0x1 Uid=1000 Gid=100 Pid=12880] "nyc.jpg"
stat {vol1 /nyc.jpg}
 -> [ID=0xb] Lookup error=ENOENT
 <- Lookup [ID=0xc Node=0x1 Uid=1000 Gid=100 Pid=12880] "nyc.jpg"
stat {vol1 /nyc.jpg}
 -> [ID=0xc] Lookup error=ENOENT
 <- Rename [ID=0xd Node=0x1 Uid=1000 Gid=100 Pid=12880] from "nyc2.jpg" to dirnode 0x1 "nyc.jpg"
renaming {/nyc2.jpg} {/nyc.jpg} 
 => InvalidateNode 0x1 Off:0 Size:-1
 -> [ID=0xd] Rename
 <- Getattr [ID=0xe Node=0x1 Uid=1000 Gid=100 Pid=12892] 0x0 fl=0
 -> [ID=0xe] Getattr valid=1m0s ino=1 size=0 mode=dr-xr-xr-x
 <- Open [ID=0xf Node=0x1 Uid=1000 Gid=100 Pid=12892] dir=true fl=OpenReadOnly+OpenDirectory+OpenNonblock
 -> [ID=0xf] Open 0x1 fl=0
 <- Read [ID=0x10 Node=0x1 Uid=1000 Gid=100 Pid=12892] 0x1 4096 @0x0 dir=true fl=0 lock=0 ffl=OpenReadOnly+OpenDirectory+OpenNonblock
returning dirents: []fuse.Dirent{fuse.Dirent{Inode:0x0, Type:0x8, Name:"nyc.jpg"}}
 -> [ID=0x10] Read 32
 <- Getattr [ID=0x11 Node=0x2 Uid=1000 Gid=100 Pid=12892] 0x0 fl=0
stat {/nyc2.jpg}
 -> [ID=0x11] Getattr error=EIO: file does not exist
 <- Read [ID=0x12 Node=0x1 Uid=1000 Gid=100 Pid=12892] 0x1 4096 @0x20 dir=true fl=0 lock=0 ffl=OpenReadOnly+OpenDirectory+OpenNonblock
 -> [ID=0x12] Read 0
 <- Release [ID=0x13 Node=0x1 Uid=0 Gid=0 Pid=0] 0x1 fl=OpenReadOnly+OpenDirectory+OpenNonblock rfl=0 owner=0x0
 -> [ID=0x13] Release
```

Where the non-bazil lines are my own sanity check. The glaring error is when it stats the file that just moved.

**Comment by tv42** [comment]
Your filesystem seems to cache filenames in its implementation, and not adjust its internal state on `Rename`.

```
 <- Lookup [ID=0x5 Node=0x1 Uid=1000 Gid=100 Pid=12861] "nyc2.jpg"
stat {/nyc2.jpg}
stat {/nyc2.jpg}
 -> [ID=0x5] Lookup 0x2 gen=0 valid=1m0s attr={valid=1m0s ino=16273471275109000053 size=1774691 mode=-rw-rw-rw-}
```

The file that is at this moment in time named `nyc2.jpg` is known as NodeID 0x2.

```
 <- Rename [ID=0xd Node=0x1 Uid=1000 Gid=100 Pid=12880] from "nyc2.jpg" to dirnode 0x1 "nyc.jpg"
renaming {/nyc2.jpg} {/nyc.jpg} 
 => InvalidateNode 0x1 Off:0 Size:-1
 -> [ID=0xd] Rename
```

Rename `nyc2.jpg` to `nyc.jpg`. It's still NodeID 0x2.

```
 <- Getattr [ID=0x11 Node=0x2 Uid=1000 Gid=100 Pid=12892] 0x0 fl=0
stat {/nyc2.jpg}
 -> [ID=0x11] Getattr error=EIO: file does not exist
```

Get attributes of NodeID 0x2 -- and your filesystem internal log talks about the old name.

This seems to be a case of misunderstanding unix filesystem structure. A file renamed retains its identity (inode number etc).

If you think there's an actual `bazil.org/fuse` bug here, please share something reproducible.

@gourao, you're the original submitter so I can't be sure if your problem is the same as @barakmich. Please provide more information.

**Comment by zhengxiaochuan-3** [comment]
+1. I'm seeing this too.  my fs keep the inodeid when rename a file .  readdirall find inodeid not change , then, fuse will cache the info until next loopup.   I guess  if inodeid change , things goes well.

**Comment by tv42** [comment]
Can't do anything here. It would help if you shared a small reproducible example -- right now, best bet is your filesystem logic is buggy.

**Comment by rkfg** [comment]
This affects me too. Here's a demo: https://gist.github.com/rkfg/db44a93dba871c74dd97e9c9b83ee92a

I do `mv mnt/test.txt mnt/test2.txt`. The output (after removing the junk from the system searching for autorun.inf etc.):
```
2020/08/09 20:44:26 Read dir all
2020/08/09 20:44:26 Lookup test.txt
2020/08/09 20:44:32 Lookup failed for test2.txt
2020/08/09 20:44:32 Lookup failed for test2.txt
2020/08/09 20:44:32 Lookup failed for test2.txt
2020/08/09 20:44:32 Renamed test.txt to test2.txt
2020/08/09 20:44:37 Read dir all
2020/08/09 20:44:37 Wrong filename, expected test2.txt got test.txt
```
So the library read the directory but didn't use it for lookup thus trying to access the non-existent file. After a minute it invalidates the cache somehow and uses lookup again and gets the actual information:
```
2020/08/09 20:46:39 Read dir all
2020/08/09 20:46:39 Lookup test2.txt
2020/08/09 20:46:39 Lookup test2.txt
```
Looks like a stale cache to me but I might do something wrong.

**Comment by rkfg** [comment]
I think the problem here is that we can't update the child node (`file`) from the parent node (`rootDir`) where rename handling happens. At the same time, `Attr` doesn't provide the filename so we should remember it internally somehow and it causes this desync. Anyway, this behavior is most probably correct and I'll use inodes as a source of truth (database IDs) in my filesystem.

**Comment by tv42** [comment]
The code in https://github.com/bazil/fuse/issues/119#issuecomment-671081429 is racy *and* has the same bug as above: "Your filesystem seems to cache filenames in its implementation, and not adjust its internal state on Rename."

The kernel has no need to Lookup something that it already has in dentry cache.

You can construct a filesystem that retains state in the returned Node and Handle instances just fine. Done that several times. You "just" need to understand the Linux VFS. And there is no way to make a Linux filesystem without understanding the Linux VFS; FUSE makes that possible in userspace, but does not decrease the rest of the learning curve.

> But we can't get the respective file while doing Rename in dir to change the internal filename 

That's a problem with your filesystem implementation. If you wish to store such state in in you Node implementation for a file, you need to be able to reach that state to update it, for example either via the parent directory or via the FS.

Also, renames across directories are a thing.

I see nothing here but "filesystems are complex". Which is not a false claim, but it's also not going to get "solved" by this library; bazil.org/fuse is a mere adapter between Go and the Linux kernel.

**Comment by rkfg** [comment]
> I see nothing here but "filesystems are complex".

That's true but it's also a bit leaky abstraction. The library provides both low and high level APIs, I don't think this problem occurs when using the low level approach. The mental model with high level node API is different, I expect it to do the things I described automatically so at every moment I have all data I need (parameters from FUSE + my own internal state) to provide a correct response. And it was true until I started implementing `Rename` which is crucial for my FS. But here the abstraction leaked, and the library used an old structure when I expected it to create a new one for me and forget the outdated one.

So yep, it's all complex, it requires some time to learn the stuff under the hood, and the library can't do anything if the kernel doesn't issue the request I expected it to issue, but documenting this particular moment wouldn't hurt. It's not obvious and confusing even though the library is not the one to blame. On the whole I like the node API, it's much easier to use for complex filesystems with different logic on different levels compared to the raw API (I wrote some FUSE filesystems in Python, Java and C++ before).
